[BUG] Android variants / dependency variants can not be resolved

**Describe the bug**
Android projects which use nested modules that are themselves full Android projects (with application variants) cause the dependency resolution to fail:
```
$ ./gradlew :app:nexusIQScan        
Configuration on demand is an incubating feature.
> Task :app:nexusIQScan FAILED
1 actionable task: 1 executed
<-------------> 0% WAITING
> :app:nexusIQScan > Resolve dependencies of :app:debugAndroidTestCompileClassp

FAILURE: Build failed with an exception.

* What went wrong:
Execution failed for task ':app:nexusIQScan'.
> Could not resolve all dependencies for configuration ':app:debugAndroidTestCompileClasspath'.
   > More than one variant of project :commonlib matches the consumer attributes:
       - Configuration ':commonlib:debugApiElements' variant android-aidl:
           - Unmatched attributes:
               - Found artifactType 'android-aidl' but wasn't required.
               - Found com.android.build.api.attributes.VariantAttr 'debug' but wasn't required.
           - Compatible attributes:
               - Required com.android.build.api.attributes.BuildTypeAttr 'debug' and found compatible value 'debug'.
               - Required com.android.build.gradle.internal.dependency.AndroidTypeAttr 'Aar' and found compatible value 'Aar'.
               - Required org.gradle.usage 'java-api' and found compatible value 'java-api'.
               - Required org.jetbrains.kotlin.platform.type 'androidJvm' and found compatible value 'androidJvm'.
       - Configuration ':commonlib:debugApiElements' variant android-manifest:
           - Unmatched attributes:
               - Found artifactType 'android-manifest' but wasn't required.
               - Found com.android.build.api.attributes.VariantAttr 'debug' but wasn't required.
           - Compatible attributes:
               - Required com.android.build.api.attributes.BuildTypeAttr 'debug' and found compatible value 'debug'.
               - Required com.android.build.gradle.internal.dependency.AndroidTypeAttr 'Aar' and found compatible value 'Aar'.
               - Required org.gradle.usage 'java-api' and found compatible value 'java-api'.
               - Required org.jetbrains.kotlin.platform.type 'androidJvm' and found compatible value 'androidJvm'.
       - Configuration ':commonlib:debugApiElements' variant android-renderscript:
           - Unmatched attributes:
               - Found artifactType 'android-renderscript' but wasn't required.
               - Found com.android.build.api.attributes.VariantAttr 'debug' but wasn't required.
           - Compatible attributes:
               - Required com.android.build.api.attributes.BuildTypeAttr 'debug' and found compatible value 'debug'.
               - Required com.android.build.gradle.internal.dependency.AndroidTypeAttr 'Aar' and found compatible value 'Aar'.
               - Required org.gradle.usage 'java-api' and found compatible value 'java-api'.
               - Required org.jetbrains.kotlin.platform.type 'androidJvm' and found compatible value 'androidJvm'.
       - Configuration ':commonlib:debugApiElements' variant android-symbol-with-package-name:
           - Unmatched attributes:
               - Found artifactType 'android-symbol-with-package-name' but wasn't required.
               - Found com.android.build.api.attributes.VariantAttr 'debug' but wasn't required.
           - Compatible attributes:
               - Required com.android.build.api.attributes.BuildTypeAttr 'debug' and found compatible value 'debug'.
               - Required com.android.build.gradle.internal.dependency.AndroidTypeAttr 'Aar' and found compatible value 'Aar'.
               - Required org.gradle.usage 'java-api' and found compatible value 'java-api'.
               - Required org.jetbrains.kotlin.platform.type 'androidJvm' and found compatible value 'androidJvm'.

* Try:
Run with --stacktrace option to get the stack trace. Run with --info or --debug option to get more log output. Run with --scan to get full insights.

* Get more help at https://help.gradle.org
```


**To Reproduce**
1. Checkout sample project: https://github.com/inventage/iqserver-android
2. Copy `app/gradle.template.properties` to `app/gradle.properties` and set nexusIQServer properties
3. Run `./gradlew :app:nexusIQScan`


**Expected behavior**
I expect the dependencies of the module `app` to be gathered and submitted for a scan.


**Desktop (please complete the following information):**
 - OS: Arch Linux
 - Gradle Version: 5.6.4 (Gradle Wrapper)
 - JVM Version and Flavor: OpenJDK, Java(TM) SE Runtime Environment (build 1.8.0_261-b12)
 - Nexus IQ Server Version: 93
 - Version 1.2.0

**Additional context**
The problem is related to Gradle's concept of dependency variants. The Android build system uses it heavily, but the problem is not exclusive to Android submodules, it also shows up when using dependencies that are published with Gradle variant metadata and having the experimental metadata feature enabled.

Not all resolvable configurations have enough metadata to determine which variants of dependencies should be selected and the ones that do in an Android project might not contain jars/aars anymore. We could not find a single existing configuration that is both *actually* resolvable and contains the aars we need to successfully scan. There is a way to extract the actual compile classpath of a single variant (see https://proandroiddev.com/how-to-extract-your-android-projects-dependencies-with-a-gradle-task-1775e8c21093 and https://gist.github.com/n-belokopytov/d44949590748b096c1a497008b761d04) that was also at one point featured on the Sonatype blog, but that yields just the contained jars of any aars that an Android build might use as dependency.

We managed to extract the full set of dependencies eventually, with our custom Gradle plugin in the referenced sample project. (see https://github.com/inventage/iqserver-android/blob/master/buildSrc/src/main/java/com/inventage/gradle/android/AndroidResolvableConfigPlugin.java)
The plugin works like this:
1. Create a new configuration in the build, that configuration is to be used to gather dependencies for the scan.
2. For all dependencies added to a set of existing configurations (including extended configurations):
  1. If the dependency is to another Gradle module, repeat this algorithm for that module, starting at step 2
  2. If the dependency is anything else, add the dependency to our new configuration
Then in the build, add a Copy task that copies the contents of our configuration to a directory we went. Externally to the build, scan the directory with the iq server CLI.
For external dependencies with variants, it is necessary to add Attributes to our configuration to help Gradle select the dependency variant. That is now hardcoded to "org.gradle.usage" -> "java-runtime", but might have to be configurable.

It would be great to see this adapted algorithm included in the Sonatype plugin.

This is probably related to https://github.com/sonatype-nexus-community/scan-gradle-plugin/issues/32 in that our method uses separate Configurations and a Copy task to copy dependencies instead of resolving them 'manually' in the plugin.

--- a/.circleci/config.yml
+++ b/.circleci/config.yml
@@ -1,10 +1,10 @@
 version: 2.1
 orbs:
-  gradle: circleci/android:api-28-alpha
+  gradle: circleci/gradle@1.0.11
 executors:
   openjdk-8-executor:
     docker:
-      - image: 'circleci/openjdk:8'
+      - image: 'circleci/android:api-28-alpha'
 
 commands:
   configure-git:

--- a/.circleci/config.yml
+++ b/.circleci/config.yml
@@ -1,6 +1,6 @@
 version: 2.1
 orbs:
-  gradle: circleci/gradle@1.0.11
+  gradle: circleci/android:api-28-alpha
 executors:
   openjdk-8-executor:
     docker:

--- a/README.md
+++ b/README.md
@@ -143,21 +143,6 @@ As mentioned above the values can be set on command line using -D arguments or i
 Just apply the plugin on the root project and all sub-modules will be processed and the output will be a single report
 with all components found in each module. This includes Android projects.
 
-### Known issue with dependencies and build variants (mostly found in Android projects)
-Let's say an Android project has a module `core` with some code (library) and another module `app` as the actual Android
-app. When using build variants an error like this may show:
-
-> Could not resolve all dependencies for configuration ':app:releaseCompileClasspath'.
-More than one variant of project :core matches the consumer attributes
-      Configuration ':core:releaseApiElements' variant android-aidl: Unmatched attributes:
-      - Found artifactType 'android-aidl' but wasn't required.
-      - Found com.android.build.api.attributes.VariantAttr 'release' but wasn't required.
-
-At the moment, to get ride of that variant issue we recommend to change the way the library is declared in `app` by
-going to the `build.gradle` file from:
-    - `api project(':core')` or `implementation project(':core')` to
-    - `api project(path: ':core', configuration: 'default')` or `implementation project(path: ':core', configuration: 'default')`
-
 ## Contributing
 
 We care a lot about making the world a safer place, and that's why we created this `scan-gradle-plugin`. If you as well want to speed up the pace of software development by working on this project, jump on in! Before you start work, create a new issue, or comment on an existing issue, to let others know you are!

--- a/build.gradle
+++ b/build.gradle
@@ -90,6 +90,7 @@ dependencies {
   testCompile "commons-io:commons-io:$commonsIoVersion"
   testCompile "org.powermock:powermock-module-junit4:$powermockVersion"
   testCompile "org.powermock:powermock-api-mockito2:$powermockVersion"
+  testCompile "com.github.ardenliu:arden-file:$ardenFileVersion"
 }
 
 processResources {

--- a/gradle.properties
+++ b/gradle.properties
@@ -25,3 +25,4 @@ junitVersion=4.13
 powermockVersion=2.0.4
 assertJVersion=3.14.0
 commonsIoVersion=2.6
+ardenFileVersion=0.0.4

--- a/gradle/integration-test.gradle
+++ b/gradle/integration-test.gradle
@@ -13,7 +13,7 @@ task integrationTest(type: Test) {
 }
 
 ['**/ScanIT_Gradle_Versions_3_3*.class',
- '**/ScanIT_Gradle_Versions_4_0*.class',
+ '**/ScanIT_Gradle_Versions_4_1*.class',
  '**/ScanIT_Gradle_Versions_4_4*.class',
  '**/ScanIT_Gradle_Versions_5_0*.class',
  '**/ScanIT_Gradle_Versions_5_5*.class',

--- a/src/integTest/java/org/sonatype/gradle/plugins/scan/ScanIT_Gradle_Versions_3_3_to_4_0.java
+++ b/src/integTest/java/org/sonatype/gradle/plugins/scan/ScanIT_Gradle_Versions_3_3_to_4_0.java
@@ -5,15 +5,15 @@ import java.util.List;
 
 import org.junit.runners.Parameterized;
 
-public class ScanIT_Gradle_Versions_3_3_to_3_5
+public class ScanIT_Gradle_Versions_3_3_to_4_0
     extends ScanPluginIntegrationTestBase
 {
   @Parameterized.Parameters(name = "Version: {0}")
   public static List<String> data() {
-    return Arrays.asList("3.3", "3.4.1", "3.5");
+    return Arrays.asList("3.3", "3.5", "4.0");
   }
 
-  public ScanIT_Gradle_Versions_3_3_to_3_5(final String gradleVersion) {
+  public ScanIT_Gradle_Versions_3_3_to_4_0(final String gradleVersion) {
     super(gradleVersion);
     this.useLegacySyntax = true;
   }

--- a/src/integTest/java/org/sonatype/gradle/plugins/scan/ScanIT_Gradle_Versions_4_1_to_4_3.java
+++ b/src/integTest/java/org/sonatype/gradle/plugins/scan/ScanIT_Gradle_Versions_4_1_to_4_3.java
@@ -5,15 +5,15 @@ import java.util.List;
 
 import org.junit.runners.Parameterized;
 
-public class ScanIT_Gradle_Versions_4_0_to_4_3
+public class ScanIT_Gradle_Versions_4_1_to_4_3
     extends ScanPluginIntegrationTestBase
 {
   @Parameterized.Parameters(name = "Version: {0}")
   public static List<String> data() {
-    return Arrays.asList("4.0", "4.1", "4.3.1");
+    return Arrays.asList("4.1", "4.2.1", "4.3.1");
   }
 
-  public ScanIT_Gradle_Versions_4_0_to_4_3(final String gradleVersion) {
+  public ScanIT_Gradle_Versions_4_1_to_4_3(final String gradleVersion) {
     super(gradleVersion);
   }
 }

--- a/src/integTest/java/org/sonatype/gradle/plugins/scan/ScanPluginIntegrationTestBase.java
+++ b/src/integTest/java/org/sonatype/gradle/plugins/scan/ScanPluginIntegrationTestBase.java
@@ -20,8 +20,10 @@ import java.io.File;
 import java.io.FileWriter;
 import java.io.IOException;
 import java.io.InputStream;
+import java.net.URISyntaxException;
 import java.nio.charset.StandardCharsets;
 
+import com.github.ardenliu.common.file.ResourcesUtils;
 import org.apache.commons.io.IOUtils;
 import org.gradle.testkit.runner.BuildResult;
 import org.gradle.testkit.runner.GradleRunner;
@@ -132,6 +134,22 @@ public abstract class ScanPluginIntegrationTestBase
   }
 
   @Test
+  public void testScanTask_Android_NexusIQ() throws IOException, InterruptedException, URISyntaxException {
+    String resource = useLegacySyntax ? "legacy-syntax" + File.separator + "android" : "android";
+    ResourcesUtils.copyFromClassPath(resource, testProjectDir.getRoot().toPath());
+
+    final BuildResult result = GradleRunner.create()
+        .withGradleVersion(gradleVersion)
+        .withProjectDir(new File(testProjectDir.getRoot(), resource))
+        .withPluginClasspath()
+        .withArguments("nexusIQScan", "--info")
+        .build();
+
+    assertBuildOutputText_NexusIQ(result, "None");
+    assertThat(result.task(":nexusIQScan").getOutcome()).isEqualTo(SUCCESS);
+  }
+
+  @Test
   public void testAuditTask_NoVulnerabilities_OssIndex() throws IOException {
     writeFile(buildFile, "control.gradle");
 
@@ -197,6 +215,28 @@ public abstract class ScanPluginIntegrationTestBase
         .withPluginClasspath().withArguments("ossIndexAudit", "--info").build();
 
     assertBuildOutputText_OssIndex(result, 0);
+    assertThat(result.task(":ossIndexAudit").getOutcome()).isEqualTo(SUCCESS);
+  }
+
+  @Test
+  public void testAuditTask_Android_OssIndex() throws IOException, InterruptedException, URISyntaxException {
+    String resource = useLegacySyntax ? "legacy-syntax" + File.separator + "android" : "android";
+    ResourcesUtils.copyFromClassPath(resource, testProjectDir.getRoot().toPath());
+
+    BuildResult result = GradleRunner.create()
+        .withGradleVersion(gradleVersion)
+        .withProjectDir(new File(testProjectDir.getRoot(), resource))
+        .withPluginClasspath()
+        .withArguments("ossIndexAudit", "--info").build();
+
+    String resultOutput = result.getOutput();
+    assertThat(resultOutput).contains(String.format(
+        "%scom.android.support:appcompat-v7:24.2.1: 0 vulnerabilities detected%n"
+            + "|    %scom.android.support:animated-vector-drawable:24.2.1: 0 vulnerabilities detected",
+        DEPENDENCY_PREFIX, DEPENDENCY_PREFIX));
+    assertThat(resultOutput).contains(
+        String.format("%scommons-collections:commons-collections:3.1: 0 vulnerabilities detected", DEPENDENCY_PREFIX));
+
     assertThat(result.task(":ossIndexAudit").getOutcome()).isEqualTo(SUCCESS);
   }
 

--- a/src/integTest/java/org/sonatype/gradle/plugins/scan/ScanPluginIntegrationTestBase.java
+++ b/src/integTest/java/org/sonatype/gradle/plugins/scan/ScanPluginIntegrationTestBase.java
@@ -20,7 +20,6 @@ import java.io.File;
 import java.io.FileWriter;
 import java.io.IOException;
 import java.io.InputStream;
-import java.net.URISyntaxException;
 import java.nio.charset.StandardCharsets;
 
 import com.github.ardenliu.common.file.ResourcesUtils;
@@ -134,7 +133,7 @@ public abstract class ScanPluginIntegrationTestBase
   }
 
   @Test
-  public void testScanTask_Android_NexusIQ() throws IOException, InterruptedException, URISyntaxException {
+  public void testScanTask_Android_NexusIQ() {
     String resource = useLegacySyntax ? "legacy-syntax" + File.separator + "android" : "android";
     ResourcesUtils.copyFromClassPath(resource, testProjectDir.getRoot().toPath());
 
@@ -200,8 +199,12 @@ public abstract class ScanPluginIntegrationTestBase
   public void testAuditTask_ExcludeTestDependencies_OssIndex() throws IOException {
     writeFile(buildFile, "exclude-test-dependency.gradle");
 
-    BuildResult result = GradleRunner.create().withGradleVersion(gradleVersion).withProjectDir(testProjectDir.getRoot())
-        .withPluginClasspath().withArguments("ossIndexAudit", "--info").build();
+    BuildResult result = GradleRunner.create()
+        .withGradleVersion(gradleVersion)
+        .withProjectDir(testProjectDir.getRoot())
+        .withPluginClasspath()
+        .withArguments("ossIndexAudit", "--info")
+        .build();
 
     assertThat(result.getOutput()).contains("0 dependencies");
     assertThat(result.task(":ossIndexAudit").getOutcome()).isEqualTo(SUCCESS);
@@ -211,15 +214,19 @@ public abstract class ScanPluginIntegrationTestBase
   public void testAuditTask_IncludeTestDependencies_OssIndex() throws IOException {
     writeFile(buildFile, "include-test-dependency.gradle");
 
-    BuildResult result = GradleRunner.create().withGradleVersion(gradleVersion).withProjectDir(testProjectDir.getRoot())
-        .withPluginClasspath().withArguments("ossIndexAudit", "--info").build();
+    BuildResult result = GradleRunner.create()
+        .withGradleVersion(gradleVersion)
+        .withProjectDir(testProjectDir.getRoot())
+        .withPluginClasspath()
+        .withArguments("ossIndexAudit", "--info")
+        .build();
 
     assertBuildOutputText_OssIndex(result, 0);
     assertThat(result.task(":ossIndexAudit").getOutcome()).isEqualTo(SUCCESS);
   }
 
   @Test
-  public void testAuditTask_Android_OssIndex() throws IOException, InterruptedException, URISyntaxException {
+  public void testAuditTask_Android_OssIndex() {
     String resource = useLegacySyntax ? "legacy-syntax" + File.separator + "android" : "android";
     ResourcesUtils.copyFromClassPath(resource, testProjectDir.getRoot().toPath());
 
@@ -227,7 +234,8 @@ public abstract class ScanPluginIntegrationTestBase
         .withGradleVersion(gradleVersion)
         .withProjectDir(new File(testProjectDir.getRoot(), resource))
         .withPluginClasspath()
-        .withArguments("ossIndexAudit", "--info").build();
+        .withArguments("ossIndexAudit", "--info")
+        .build();
 
     String resultOutput = result.getOutput();
     assertThat(resultOutput).contains(String.format(

--- a/src/integTest/resources/android/app/build.gradle
+++ b/src/integTest/resources/android/app/build.gradle
@@ -1,0 +1,19 @@
+apply plugin: 'com.android.application'
+
+android {
+  compileSdkVersion rootProject.ext.compileSdkVersion
+
+  defaultConfig {
+    applicationId 'org.sonatype.gradle.plugins.scan.android'
+    minSdkVersion rootProject.ext.minSdkVersion
+    targetSdkVersion rootProject.ext.targetSdkVersion
+    versionCode 1
+    versionName '1.0.0'
+    buildToolsVersion rootProject.ext.buildToolsVersion
+  }
+}
+
+dependencies {
+  implementation 'com.android.support:appcompat-v7:24.2.1'
+  implementation project(path: ':common-lib')
+}

--- a/src/integTest/resources/android/app/src/main/AndroidManifest.xml
+++ b/src/integTest/resources/android/app/src/main/AndroidManifest.xml
@@ -1,0 +1,5 @@
+<manifest package="com.inventage.sonatype.android.app"
+          xmlns:android="http://schemas.android.com/apk/res/android"
+          xmlns:tools="http://schemas.android.com/tools">
+  <application android:name="org.sonatype.gradle.plugins.scan.android" />
+</manifest>

--- a/src/integTest/resources/android/build.gradle
+++ b/src/integTest/resources/android/build.gradle
@@ -23,9 +23,9 @@ allprojects {
 }
 
 ext {
-  compileSdkVersion=29
+  compileSdkVersion=28
   minSdkVersion=10
-  targetSdkVersion=29
+  targetSdkVersion=28
   buildToolsVersion="26.0.2"
 }
 

--- a/src/integTest/resources/android/build.gradle
+++ b/src/integTest/resources/android/build.gradle
@@ -1,0 +1,42 @@
+buildscript {
+  repositories {
+    google()
+    mavenCentral()
+    jcenter()
+  }
+
+  dependencies {
+    classpath 'com.android.tools.build:gradle:3.0.0'
+  }
+}
+
+plugins {
+  id 'org.sonatype.gradle.plugins.scan'
+}
+
+allprojects {
+  repositories {
+    google()
+    mavenCentral()
+    jcenter()
+  }
+}
+
+ext {
+  compileSdkVersion=29
+  minSdkVersion=10
+  targetSdkVersion=29
+  buildToolsVersion="26.0.2"
+}
+
+nexusIQScan {
+  serverUrl = 'http://localhost:8081'
+  username = 'admin'
+  password = 'admin123'
+  applicationId = 'testing-gradle-plugin'
+  simulationEnabled = true
+}
+
+ossIndexAudit {
+  simulationEnabled = true
+}

--- a/src/integTest/resources/android/build.gradle
+++ b/src/integTest/resources/android/build.gradle
@@ -5,8 +5,17 @@ buildscript {
     jcenter()
   }
 
+  // Mapping between Gradle and Android plugin versions: https://developer.android.com/studio/releases/gradle-plugin
+  def androidGradlePluginVersion = '3.0.0'
+  if (gradle.gradleVersion.compareTo('6.1.1') >= 0) {
+    androidGradlePluginVersion = '4.0.0'
+  } else if (gradle.gradleVersion.compareTo('5.6.4') >= 0) {
+    androidGradlePluginVersion = '3.6.4'
+  } else if (gradle.gradleVersion.compareTo('5.0') >= 0) {
+    androidGradlePluginVersion = '3.3.3'
+  }
   dependencies {
-    classpath 'com.android.tools.build:gradle:3.0.0'
+    classpath "com.android.tools.build:gradle:$androidGradlePluginVersion"
   }
 }
 
@@ -26,7 +35,7 @@ ext {
   compileSdkVersion=28
   minSdkVersion=10
   targetSdkVersion=28
-  buildToolsVersion="26.0.2"
+  buildToolsVersion='26.0.2'
 }
 
 nexusIQScan {

--- a/src/integTest/resources/android/common-lib/build.gradle
+++ b/src/integTest/resources/android/common-lib/build.gradle
@@ -1,0 +1,17 @@
+apply plugin: 'com.android.library'
+
+android {
+  compileSdkVersion rootProject.ext.compileSdkVersion
+
+  defaultConfig {
+    minSdkVersion rootProject.ext.minSdkVersion
+    targetSdkVersion rootProject.ext.targetSdkVersion
+    versionCode 1
+    versionName '1.0'
+    buildToolsVersion rootProject.ext.buildToolsVersion
+  }
+}
+
+dependencies {
+  implementation 'commons-collections:commons-collections:3.1'
+}

--- a/src/integTest/resources/android/common-lib/src/main/AndroidManifest.xml
+++ b/src/integTest/resources/android/common-lib/src/main/AndroidManifest.xml
@@ -1,0 +1,5 @@
+<manifest package="com.inventage.sonatype.android.app"
+          xmlns:android="http://schemas.android.com/apk/res/android"
+          xmlns:tools="http://schemas.android.com/tools">
+  <application android:name="org.sonatype.gradle.plugins.scan.android" />
+</manifest>

--- a/src/integTest/resources/android/settings.gradle
+++ b/src/integTest/resources/android/settings.gradle
@@ -1,0 +1,1 @@
+include ':common-lib', ':app'

--- a/src/integTest/resources/legacy-syntax/android/app/build.gradle
+++ b/src/integTest/resources/legacy-syntax/android/app/build.gradle
@@ -1,0 +1,19 @@
+apply plugin: 'com.android.application'
+
+android {
+  compileSdkVersion rootProject.ext.compileSdkVersion
+
+  defaultConfig {
+    applicationId 'org.sonatype.gradle.plugins.scan.android'
+    minSdkVersion rootProject.ext.minSdkVersion
+    targetSdkVersion rootProject.ext.targetSdkVersion
+    versionCode 1
+    versionName '1.0.0'
+    buildToolsVersion rootProject.ext.buildToolsVersion
+  }
+}
+
+dependencies {
+  compile 'com.android.support:appcompat-v7:24.2.1'
+  compile project(path: ':common-lib')
+}

--- a/src/integTest/resources/legacy-syntax/android/app/src/main/AndroidManifest.xml
+++ b/src/integTest/resources/legacy-syntax/android/app/src/main/AndroidManifest.xml
@@ -1,0 +1,5 @@
+<manifest package="com.inventage.sonatype.android.app"
+          xmlns:android="http://schemas.android.com/apk/res/android"
+          xmlns:tools="http://schemas.android.com/tools">
+  <application android:name="org.sonatype.gradle.plugins.scan.android" />
+</manifest>

--- a/src/integTest/resources/legacy-syntax/android/build.gradle
+++ b/src/integTest/resources/legacy-syntax/android/build.gradle
@@ -21,9 +21,9 @@ allprojects {
 }
 
 ext {
-  compileSdkVersion=29
+  compileSdkVersion=28
   minSdkVersion=10
-  targetSdkVersion=29
+  targetSdkVersion=28
   buildToolsVersion="26.0.2"
 }
 

--- a/src/integTest/resources/legacy-syntax/android/build.gradle
+++ b/src/integTest/resources/legacy-syntax/android/build.gradle
@@ -24,7 +24,7 @@ ext {
   compileSdkVersion=28
   minSdkVersion=10
   targetSdkVersion=28
-  buildToolsVersion="26.0.2"
+  buildToolsVersion='26.0.2'
 }
 
 nexusIQScan {

--- a/src/integTest/resources/legacy-syntax/android/build.gradle
+++ b/src/integTest/resources/legacy-syntax/android/build.gradle
@@ -1,0 +1,40 @@
+buildscript {
+  repositories {
+    maven { url 'https://maven.google.com' }
+    mavenCentral()
+  }
+
+  dependencies {
+    classpath 'com.android.tools.build:gradle:2.3.0'
+  }
+}
+
+plugins {
+  id 'org.sonatype.gradle.plugins.scan'
+}
+
+allprojects {
+  repositories {
+    maven { url 'https://maven.google.com' }
+    mavenCentral()
+  }
+}
+
+ext {
+  compileSdkVersion=29
+  minSdkVersion=10
+  targetSdkVersion=29
+  buildToolsVersion="26.0.2"
+}
+
+nexusIQScan {
+  serverUrl = 'http://localhost:8081'
+  username = 'admin'
+  password = 'admin123'
+  applicationId = 'testing-gradle-plugin'
+  simulationEnabled = true
+}
+
+ossIndexAudit {
+  simulationEnabled = true
+}

--- a/src/integTest/resources/legacy-syntax/android/common-lib/build.gradle
+++ b/src/integTest/resources/legacy-syntax/android/common-lib/build.gradle
@@ -1,0 +1,17 @@
+apply plugin: 'com.android.library'
+
+android {
+  compileSdkVersion rootProject.ext.compileSdkVersion
+
+  defaultConfig {
+    minSdkVersion rootProject.ext.minSdkVersion
+    targetSdkVersion rootProject.ext.targetSdkVersion
+    versionCode 1
+    versionName '1.0'
+    buildToolsVersion rootProject.ext.buildToolsVersion
+  }
+}
+
+dependencies {
+  compile 'commons-collections:commons-collections:3.1'
+}

--- a/src/integTest/resources/legacy-syntax/android/common-lib/src/main/AndroidManifest.xml
+++ b/src/integTest/resources/legacy-syntax/android/common-lib/src/main/AndroidManifest.xml
@@ -1,0 +1,5 @@
+<manifest package="com.inventage.sonatype.android.app"
+          xmlns:android="http://schemas.android.com/apk/res/android"
+          xmlns:tools="http://schemas.android.com/tools">
+  <application android:name="org.sonatype.gradle.plugins.scan.android" />
+</manifest>

--- a/src/integTest/resources/legacy-syntax/android/settings.gradle
+++ b/src/integTest/resources/legacy-syntax/android/settings.gradle
@@ -1,0 +1,1 @@
+include ':common-lib', ':app'

--- a/src/main/java/org/sonatype/gradle/plugins/scan/common/DependenciesFinder.java
+++ b/src/main/java/org/sonatype/gradle/plugins/scan/common/DependenciesFinder.java
@@ -21,7 +21,9 @@ import java.util.HashSet;
 import java.util.LinkedHashSet;
 import java.util.List;
 import java.util.Set;
+import java.util.function.Function;
 import java.util.stream.Collectors;
+import java.util.stream.Stream;
 
 import com.sonatype.insight.scan.module.model.Artifact;
 import com.sonatype.insight.scan.module.model.Module;
@@ -29,31 +31,50 @@ import com.sonatype.insight.scan.module.model.Module;
 import org.apache.commons.lang3.StringUtils;
 import org.gradle.api.Project;
 import org.gradle.api.artifacts.Configuration;
+import org.gradle.api.artifacts.ProjectDependency;
+import org.gradle.api.artifacts.ResolveException;
 import org.gradle.api.artifacts.ResolvedArtifact;
+import org.gradle.api.artifacts.ResolvedConfiguration;
 import org.gradle.api.artifacts.ResolvedDependency;
+import org.gradle.api.attributes.Usage;
+import org.gradle.api.model.ObjectFactory;
 import org.gradle.internal.impldep.com.google.common.annotations.VisibleForTesting;
 
 import static org.gradle.api.plugins.JavaPlugin.COMPILE_CLASSPATH_CONFIGURATION_NAME;
 
 public class DependenciesFinder
 {
+  private static final String RELEASE_COMPILE_LEGACY_CONFIGURATION_NAME = "_releaseCompile";
+
   private static final String RELEASE_COMPILE_CONFIGURATION_NAME = "releaseCompileClasspath";
 
   private static final Set<String> CONFIGURATION_NAMES =
-      new HashSet<>(Arrays.asList(COMPILE_CLASSPATH_CONFIGURATION_NAME, RELEASE_COMPILE_CONFIGURATION_NAME));
+      new HashSet<>(Arrays.asList(
+          COMPILE_CLASSPATH_CONFIGURATION_NAME,
+          RELEASE_COMPILE_LEGACY_CONFIGURATION_NAME,
+          RELEASE_COMPILE_CONFIGURATION_NAME));
+
+  private static final String COPY_CONFIGURATION_NAME = "sonatypeCopyConfiguration";
+
+  private static final String ATTRIBUTES_SUPPORTED_GRADLE_VERSION = "4.0";
 
   public Set<ResolvedDependency> findResolvedDependencies(Project rootProject, boolean allConfigurations) {
-    return rootProject.getAllprojects().stream()
-        .flatMap(project -> project.getConfigurations().stream())
+    Configuration copyConfiguration = createCopyConfiguration(rootProject);
+
+    return rootProject.getAllprojects().stream().flatMap(project -> project.getConfigurations().stream())
         .filter(configuration -> isAcceptableConfiguration(configuration, allConfigurations))
-        .flatMap(configuration -> configuration.getResolvedConfiguration().getFirstLevelModuleDependencies().stream())
+        .flatMap(configuration -> getDependencies(rootProject, configuration, copyConfiguration,
+            resolvedConfiguration -> resolvedConfiguration.getFirstLevelModuleDependencies().stream()))
         .collect(Collectors.toCollection(LinkedHashSet::new));
   }
 
   public Set<ResolvedArtifact> findResolvedArtifacts(Project project, boolean allConfigurations) {
+    Configuration copyConfiguration = createCopyConfiguration(project);
+
     return project.getConfigurations().stream()
         .filter(configuration -> isAcceptableConfiguration(configuration, allConfigurations))
-        .flatMap(configuration -> configuration.getResolvedConfiguration().getResolvedArtifacts().stream())
+        .flatMap(configuration -> getDependencies(project, configuration, copyConfiguration,
+            resolvedConfiguration -> resolvedConfiguration.getResolvedArtifacts().stream()))
         .collect(Collectors.toSet());
   }
 
@@ -64,10 +85,8 @@ public class DependenciesFinder
       Module module = buildModule(project);
 
       findResolvedArtifacts(project, allConfigurations).forEach(resolvedArtifact -> {
-        Artifact artifact = new Artifact()
-            .setId(resolvedArtifact.getId().getComponentIdentifier().getDisplayName())
-            .setPathname(resolvedArtifact.getFile())
-            .setMonitored(true);
+        Artifact artifact = new Artifact().setId(resolvedArtifact.getId().getComponentIdentifier().getDisplayName())
+            .setPathname(resolvedArtifact.getFile()).setMonitored(true);
         module.addConsumedArtifact(artifact);
       });
 
@@ -79,9 +98,7 @@ public class DependenciesFinder
 
   @VisibleForTesting
   Module buildModule(Project project) {
-    Module module = new Module()
-        .setIdKind("gradle")
-        .setPathname(project.getProjectDir());
+    Module module = new Module().setIdKind("gradle").setPathname(project.getProjectDir());
 
     StringBuilder idBuilder = new StringBuilder();
     if (StringUtils.isNotBlank(project.getGroup().toString())) {
@@ -97,11 +114,55 @@ public class DependenciesFinder
     return module;
   }
 
+  @VisibleForTesting
+  Configuration createCopyConfiguration(Project project) {
+    Configuration copyConfiguration = project.getConfigurations().maybeCreate(COPY_CONFIGURATION_NAME);
+    if (isGradleVersionSupportedForAttributes(project.getGradle().getGradleVersion())) {
+      copyConfiguration.attributes(attributeContainer -> {
+        ObjectFactory factory = project.getObjects();
+        attributeContainer.attribute(Usage.USAGE_ATTRIBUTE, factory.named(Usage.class, Usage.JAVA_RUNTIME));
+      });
+    }
+    return copyConfiguration;
+  }
+
+  @VisibleForTesting
+  boolean isGradleVersionSupportedForAttributes(String gradleVersion) {
+    return gradleVersion.compareTo(ATTRIBUTES_SUPPORTED_GRADLE_VERSION) >= 0;
+  }
+
+  @VisibleForTesting
+  <T> Stream<T> getDependencies(
+      Project project,
+      Configuration originalConfiguration,
+      Configuration copyConfiguration,
+      Function<ResolvedConfiguration, Stream<T>> function)
+  {
+    try {
+      return function.apply(originalConfiguration.getResolvedConfiguration());
+    }
+    catch (ResolveException e) {
+      originalConfiguration.getAllDependencies().all(dependency -> {
+        if (dependency instanceof ProjectDependency) {
+          Project dependencyProject = ((ProjectDependency) dependency).getDependencyProject();
+          project.evaluationDependsOn(dependencyProject.getPath());
+        }
+        else {
+          copyConfiguration.getDependencies().add(dependency);
+        }
+      });
+      return function.apply(copyConfiguration.getResolvedConfiguration());
+    }
+  }
+
   private boolean isAcceptableConfiguration(Configuration configuration, boolean allConfigurations) {
+    if (configuration.getName().endsWith(COPY_CONFIGURATION_NAME))
+      return false;
     if (allConfigurations) {
       return configuration.isCanBeResolved();
     }
-    return CONFIGURATION_NAMES.contains(configuration.getName())
-        || StringUtils.endsWithIgnoreCase(configuration.getName(), RELEASE_COMPILE_CONFIGURATION_NAME);
+    return configuration.isCanBeResolved() && (CONFIGURATION_NAMES.contains(configuration.getName())
+        || StringUtils.endsWithIgnoreCase(configuration.getName(), RELEASE_COMPILE_LEGACY_CONFIGURATION_NAME)
+        || StringUtils.endsWithIgnoreCase(configuration.getName(), RELEASE_COMPILE_CONFIGURATION_NAME));
   }
 }

--- a/src/main/java/org/sonatype/gradle/plugins/scan/nexus/iq/NexusIqScanTask.java
+++ b/src/main/java/org/sonatype/gradle/plugins/scan/nexus/iq/NexusIqScanTask.java
@@ -75,6 +75,8 @@ public class NexusIqScanTask
               Collections.singletonList(new Action(extension.getSimulatedPolicyActionId()))));
         }
 
+        dependenciesFinder.findModules(getProject(), extension.isAllConfigurations());
+
         applicationPolicyEvaluation =
             new ApplicationPolicyEvaluation(0, 0, 0, 0, 0, 0, 0, 0, alerts, "simulated/report");
       }

--- a/src/test/java/org/sonatype/gradle/plugins/scan/common/DependenciesFinderTest.java
+++ b/src/test/java/org/sonatype/gradle/plugins/scan/common/DependenciesFinderTest.java
@@ -17,14 +17,19 @@ package org.sonatype.gradle.plugins.scan.common;
 
 import java.util.List;
 import java.util.Set;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
 
 import com.sonatype.insight.scan.module.model.Artifact;
 import com.sonatype.insight.scan.module.model.Module;
 
 import org.gradle.api.Project;
+import org.gradle.api.artifacts.Configuration;
+import org.gradle.api.artifacts.ResolveException;
 import org.gradle.api.artifacts.ResolvedArtifact;
 import org.gradle.api.artifacts.ResolvedDependency;
 import org.gradle.api.artifacts.dsl.DependencyHandler;
+import org.gradle.api.attributes.Usage;
 import org.gradle.testfixtures.ProjectBuilder;
 import org.junit.Before;
 import org.junit.Test;
@@ -32,6 +37,8 @@ import org.junit.Test;
 import static org.assertj.core.api.Assertions.assertThat;
 import static org.gradle.api.plugins.JavaPlugin.COMPILE_CLASSPATH_CONFIGURATION_NAME;
 import static org.gradle.api.plugins.JavaPlugin.TEST_COMPILE_CLASSPATH_CONFIGURATION_NAME;
+import static org.mockito.Mockito.mock;
+import static org.mockito.Mockito.when;
 
 public class DependenciesFinderTest
 {
@@ -52,8 +59,22 @@ public class DependenciesFinderTest
   }
 
   @Test
+  public void testFindResolvedDependencies_includeLegacyAndroidDependencies() {
+    Project project = buildProject("_releaseCompile", true);
+    Set<ResolvedDependency> result = finder.findResolvedDependencies(project, false);
+    assertThat(result).hasSize(1);
+  }
+
+  @Test
   public void testFindResolvedDependencies_includeAndroidDependencies() {
     Project project = buildProject("releaseCompileClasspath", true);
+    Set<ResolvedDependency> result = finder.findResolvedDependencies(project, false);
+    assertThat(result).hasSize(1);
+  }
+
+  @Test
+  public void testFindResolvedDependencies_includeLegacyAndroidDependenciesUsingVariant() {
+    Project project = buildProject("variantProd_ReleaseCompile", true);
     Set<ResolvedDependency> result = finder.findResolvedDependencies(project, false);
     assertThat(result).hasSize(1);
   }
@@ -172,6 +193,112 @@ public class DependenciesFinderTest
 
     Artifact artifact = childModule.getConsumedArtifacts().get(0);
     assertThat(artifact.getId()).isEqualTo(COMMONS_COLLECTIONS_DEPENDENCY);
+  }
+
+  @Test
+  public void testCreateCopyConfiguration() {
+    Project project = buildProject(COMPILE_CLASSPATH_CONFIGURATION_NAME, false);
+    Configuration configuration = finder.createCopyConfiguration(project);
+
+    assertThat(configuration).isNotNull();
+    assertThat(configuration.getName()).isEqualTo("sonatypeCopyConfiguration");
+
+    Usage expectedUsage = project.getObjects().named(Usage.class, Usage.JAVA_RUNTIME);
+    assertThat(configuration.getAttributes().getAttribute(Usage.USAGE_ATTRIBUTE)).isEqualTo(expectedUsage);
+  }
+
+  @Test
+  public void testIsGradleVersionSupportedForAttributes() {
+    assertThat(finder.isGradleVersionSupportedForAttributes("3.5")).isFalse();
+    assertThat(finder.isGradleVersionSupportedForAttributes("3.5.1")).isFalse();
+    assertThat(finder.isGradleVersionSupportedForAttributes("4.0")).isTrue();
+    assertThat(finder.isGradleVersionSupportedForAttributes("4.0.1")).isTrue();
+    assertThat(finder.isGradleVersionSupportedForAttributes("4.0.2")).isTrue();
+    assertThat(finder.isGradleVersionSupportedForAttributes("4.1")).isTrue();
+  }
+
+  @Test
+  public void testGetDependencies_ModuleDependencies() {
+    Project project = buildProject(COMPILE_CLASSPATH_CONFIGURATION_NAME, false);
+    Configuration originalConfiguration = project.getConfigurations().getByName(COMPILE_CLASSPATH_CONFIGURATION_NAME);
+    Configuration copyConfiguration = finder.createCopyConfiguration(project);
+
+    Stream<ResolvedDependency> dependencies = finder.getDependencies(project, originalConfiguration, copyConfiguration,
+        resolvedConfiguration -> resolvedConfiguration.getFirstLevelModuleDependencies().stream());
+
+    assertThat(dependencies).isNotNull();
+
+    List<ResolvedDependency> list = dependencies.collect(Collectors.toList());
+    assertThat(list).hasSize(1);
+
+    Set<ResolvedArtifact> artifacts = list.get(0).getAllModuleArtifacts();
+    assertThat(artifacts).hasSize(1);
+    assertThat(artifacts.iterator().next().getId().getComponentIdentifier().toString())
+        .isEqualTo(COMMONS_COLLECTIONS_DEPENDENCY);
+  }
+
+  @Test
+  public void testGetDependencies_ModuleDependencies_WithError() {
+    Project project = buildProject(COMPILE_CLASSPATH_CONFIGURATION_NAME, false);
+
+    Configuration originalConfiguration = mock(Configuration.class);
+    when(originalConfiguration.getResolvedConfiguration()).thenThrow(ResolveException.class);
+    when(originalConfiguration.getAllDependencies())
+        .thenReturn(project.getConfigurations().getByName(COMPILE_CLASSPATH_CONFIGURATION_NAME).getAllDependencies());
+
+    Configuration copyConfiguration = finder.createCopyConfiguration(project);
+
+    Stream<ResolvedDependency> dependencies = finder.getDependencies(project, originalConfiguration, copyConfiguration,
+        resolvedConfiguration -> resolvedConfiguration.getFirstLevelModuleDependencies().stream());
+
+    assertThat(dependencies).isNotNull();
+
+    List<ResolvedDependency> list = dependencies.collect(Collectors.toList());
+    assertThat(list).hasSize(1);
+
+    Set<ResolvedArtifact> artifacts = list.get(0).getAllModuleArtifacts();
+    assertThat(artifacts).hasSize(1);
+    assertThat(artifacts.iterator().next().getId().getComponentIdentifier().toString())
+        .isEqualTo(COMMONS_COLLECTIONS_DEPENDENCY);
+  }
+
+  @Test
+  public void testGetDependencies_ResolvedArtifacts() {
+    Project project = buildProject(COMPILE_CLASSPATH_CONFIGURATION_NAME, false);
+    Configuration originalConfiguration = project.getConfigurations().getByName(COMPILE_CLASSPATH_CONFIGURATION_NAME);
+    Configuration copyConfiguration = finder.createCopyConfiguration(project);
+
+    Stream<ResolvedArtifact> dependencies = finder.getDependencies(project, originalConfiguration, copyConfiguration,
+        resolvedConfiguration -> resolvedConfiguration.getResolvedArtifacts().stream());
+
+    assertThat(dependencies).isNotNull();
+
+    List<ResolvedArtifact> list = dependencies.collect(Collectors.toList());
+
+    assertThat(list).hasSize(1);
+    assertThat(list.get(0).getId().getComponentIdentifier().toString()).isEqualTo(COMMONS_COLLECTIONS_DEPENDENCY);
+  }
+
+  @Test
+  public void testGetDependencies_ResolvedArtifacts_WithError() {
+    Project project = buildProject(COMPILE_CLASSPATH_CONFIGURATION_NAME, false);
+
+    Configuration originalConfiguration = mock(Configuration.class);
+    when(originalConfiguration.getResolvedConfiguration()).thenThrow(ResolveException.class);
+    when(originalConfiguration.getAllDependencies())
+        .thenReturn(project.getConfigurations().getByName(COMPILE_CLASSPATH_CONFIGURATION_NAME).getAllDependencies());
+
+    Configuration copyConfiguration = finder.createCopyConfiguration(project);
+
+    Stream<ResolvedArtifact> dependencies = finder.getDependencies(project, originalConfiguration, copyConfiguration,
+        resolvedConfiguration -> resolvedConfiguration.getResolvedArtifacts().stream());
+
+    assertThat(dependencies).isNotNull();
+
+    List<ResolvedArtifact> list = dependencies.collect(Collectors.toList());
+
+    assertThat(list).hasSize(1);
+    assertThat(list.get(0).getId().getComponentIdentifier().toString()).isEqualTo(COMMONS_COLLECTIONS_DEPENDENCY);
   }
 
   private Project buildProject(String configurationName, boolean needToCreateConfiguration) {

